WMF Framework/Event Platform/EventPlatformClient.swift

```diff
@@ -383,26 +353,56 @@ public class EventPlatformClient: NSObject, SamplingControllerDelegate {
      * fire-and-forget fashion
      */
     private func postAllScheduled(_ completion: (() -> Void)? = nil) {
-        DDLogDebug("EPC: Posting all scheduled requests")
+        guard let storageManager = self.storageManager else {
+            return
+        }
+
+        DDLogDebug("EPC: Processing all scheduled requests")
+        let records = storageManager.popAll()
+        if records.count == 0 {
+            DDLogDebug("EPC: Nothing to send.")
+            return
+        }
+
         let group = DispatchGroup()
-        while let item = outputBufferPopFirst() {
+        for record in records {
             group.enter()
-            httpPost(url: item.url, body: item.data) { result in
+
```

Comment:
We'll need to wrap up any Core Data object property accessors (calls to `record.stream` and `record.data` in this method) in a `performAndWait` call from `StorageManager`, to prevent threading corruption. My thinking is make a public method on StorageManager like this:

```
    func streamAndData(record: EPEventRecord) -> (stream: String, data: Data)? {
        var result: (stream: String, data: Data)? = nil
        performAndWait { moc in
            result = (record.stream, data: record.data)
        }
        return result
    }
```

Then here pull these values like so:

```
            guard let streamAndData = storageManager.streamAndData(record: record) else {
                DDLogError("EPC: Unable to extract stream and data from EPEventRecord.")
                storageManager.markFailed(record: record)
                continue
            }
            
            let streamRaw = streamAndData.stream
            let data = streamAndData.data
```

And then use `streamRaw` and `data` in place of `record.stream` and `record.data` below.

I'm pretty sure this will be enough to plug the holes, but we can try adding the [concurrency debug launch argument](https://oleb.net/blog/2014/06/core-data-concurrency-debugging/) while we test to help find any other issues.

Reply:
I was thinking (and Jason and I were talking) about how I wasn't super happy about `EPEventRecord` objects getting returned to the client. This comment makes me even less happy about it!  

Would it work to update StorageManager.popAll() to return (say) `PersistableEvent` objects containing the relevant data values, rather than the raw `EPEventRecord` objects?  Something like:

```
    func popAll() -> [EventPlatformClient.PersistableEvent] {
        var result: [EventPlatformClient.PersistableEvent] = []
        performAndWait { moc in
            let fetch: NSFetchRequest<EPEventRecord> = EPEventRecord.fetchRequest()
            fetch.sortDescriptors = [NSSortDescriptor(keyPath: \EPEventRecord.recorded, ascending: true)]
            fetch.predicate = NSPredicate(format: "(posted == nil) AND (failed != TRUE)")

            do {
                var count = 0
                let records = try moc.fetch(fetch)
                for record in records {
                    guard let stream = EventPlatformClient.Stream(rawValue: record.stream) else {
                        continue
                    }
                    result.append(EventPlatformClient.PersistableEvent(data: record.data, stream: stream))
                    count += 1
                }
                DDLogDebug("EPC: Found \(count) events awaiting submission")
            } catch let error {
                DDLogError(error.localizedDescription)
            }
        }
        return result
    }
```

Then all of our references to the `EPEventRecord` objects stay within the initial `performAndWait` block for the fetch, which (I think/hope) would resolve the threading concerns.  `PersistableEvent` would have to be updated with fields for the rest of the DB columns (which I was thinking of doing anyway) and could also perhaps take an optional `moid` field for when we want to refer back to a DB record for update based on the result of a submission attempt.

Reply:
@mdholloway yep that sounds good with using `PersistableEvent` like that - I like that it would keep full Core Data records out of `EventLoggingService`. For referring back to the DB record, your `moid` field will work, and for referencing to mark as posted or failed, I suggest using the `existingObject(with: NSManagedObjectID)` method. (https://developer.apple.com/documentation/coredata/nsmanagedobjectcontext/1506686-existingobject). Note it will throw an error if that record has since been deleted which seems possible, vs `object(with: NSManagedObjectID)` which throws an exception.

Reply:
OK, this is done in the latest version (using `existingObject`). The implementation is still a bit rough and not quite DRY but I think the idea is there at least.